funkwhale: init at 1.2.1

mweinelt opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
A social platform to enjoy and share music
https://funkwhale.audio

Syncing with and streaming to with mobile devices and activitypub sharing of music libraries.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

mweinelt committed:
funkwhale: init at 1.2.1

mweinelt commented on pkgs/servers/web-apps/funkwhale/default.nix:
```diff
@@ -0,0 +1,158 @@
+    cp -r $src/api $out/lib/funkwhale
+
+    chmod +w $out/lib/funkwhale
+    cp -r ${frontend}/ $out/lib/funkwhale/front
```
This breaks eval tests 🤷

mweinelt replied:
Looks like IFD to me.

```
error: cannot read '/build/store/qj51dlf6gibab5hm4a6h22sgicsq42zf-source/front/package.json', since path '/build/store/3khl9w0x8hx0v034p7ypsj82ygv66761-source.drv' is not valid

       at /nix/store/pyzjdbyr1shws1hhy28h94gs4drimlc3-source/lib/trivial.nix:282:24:

          281|   importJSON = path:
          282|     builtins.fromJSON (builtins.readFile path);
             |                        ^
          283|

       … while evaluating 'importJSON'
```

mweinelt commented:
IFD fixed and we have ourselves an update script to take care of all these nasty hashes.

Thanks for your help @piegamesde.

mweinelt committed:
nixos/funkwhale: init

mweinelt committed:
nixos/tests/funkwhale: init

mweinelt commented:
Today I got all services to run their respective executable and not instacrash, primarily by pinning two dependencies and iterating on a simple nixos test that requires them.

I also removed the frontend from the package output and only expose it via `funkwhale.frontend` now, which can be neatly referenced in the nginx vhosts `root`.

The module options are working quite nicely already. I have to test out what happens when the superuser information gets changed, since the setup script will call `createsuperuser` again.

One thing about the nixos test bugs me: The server component tries to connect somewhere, I don't yet know where, and I wonder if we want to patch that out :)

That's all for tonight.

mweinelt commented:
It tries to fetch files from the frontend, so it requires a complete reverse proxying setup. Stay tuned.

Visible after setting `FUNKWHALE_HOSTNAME` to `localhost`, so I could evade the dns lookup.

```pytb
 2022-01-08 04:50:38,924 django.request ERROR    Internal Server Error: /
 Traceback (most recent call last):
   File "/nix/store/fz0d0nq9x3ng0d9r6jm8spdzq1iy631v-python3.9-Django-3.2.9/lib/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
     response = get_response(request)
   File "/nix/store/g856rbdpppa0wjw96p3b833azpa98pqz-funkwhale-1.2.0/lib/funkwhale/funkwhale_api/common/middleware.py", line 239, in __call__
     return serve_spa(request)
   File "/nix/store/g856rbdpppa0wjw96p3b833azpa98pqz-funkwhale-1.2.0/lib/funkwhale/funkwhale_api/common/middleware.py", line 38, in serve_spa
     html = get_spa_html(settings.FUNKWHALE_SPA_HTML_ROOT)
   File "/nix/store/g856rbdpppa0wjw96p3b833azpa98pqz-funkwhale-1.2.0/lib/funkwhale/funkwhale_api/common/middleware.py", line 112, in get_spa_html
     return get_spa_file(spa_url, "index.html")
   File "/nix/store/g856rbdpppa0wjw96p3b833azpa98pqz-funkwhale-1.2.0/lib/funkwhale/funkwhale_api/common/middleware.py", line 130, in get_spa_file
     response.raise_for_status()
   File "/nix/store/47w9wdpsz6arsd5zf7sic5d4sj7y5gkr-python3.9-requests-2.26.0/lib/python3.9/site-packages/requests/models.py", line 953, in raise_for_status
     raise HTTPError(http_error_msg, response=self)
 requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http://localhost/front/index.html
 ```

mweinelt committed:
nixos/funkwhale: add to release notes

mweinelt committed:
nixos/nginx: Allow clientMaxBodySize on location blocks

mweinelt commented:
Today I did the nginx config with help from @fleaz.

DynamicUser and peer based authentication seems problematic. It was working for me all while I was testing and then suddenly it failed to find the user matching the UID. Very odd.

Also we can't create the superuser multiple times, so we now touch `~/.${superUserName}_created` in the state dir. That way the superuser creation only works on the initial setup. Pretty disappointing, given that paperless-ng has a `manage_superuser` option.

Learnt that we need to expose `manage.py`, to allow users to import music from the disk. Pretty disappointing workflow. Involves copying a UUID from the URL while browsing an empty library, created for the purpose of importing.

Aaaand Permissions get messy fast between funkwhale, nginx and my own user :disappointed: 

The UX of the webclient seriously needs some love. Finding things like your own libraries is annoying.

All in all this seems ready to test, I might have some local changes on my laptop still, so maybe tomorrow.

mweinelt commented on pkgs/servers/web-apps/funkwhale/update.sh:
```diff
@@ -0,0 +1,40 @@
+  const new_offline_cache_sha256 = $(nix-build -A funkwhale 2>&1 | \
+    tail -n 16 | \
+    head -n 1 | \
+    sd '\s+got:\s+' '')
```
Use a `prefetch-yarn-deps` invocation instead of this unholy mess.

mweinelt replied:
Still looking for inspiration here.

piegamesde replied:
Am I correct that this greps for a hash mismatch in lack of proper prefetching? Maybe improve the regex so that at least we don't need to match exact line positions? (something something `got:.*\nexpected:.*`, but I hate regex)

mweinelt replied:
I want to use prefetch-yarn-deps if possible. Grepping the hash mismatch is a bad joke.

mweinelt replied:
Resolved.

mweinelt commented:
Ready for review and further testing. @aanderse @davidak @piegamesde @andir

mweinelt commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,378 @@
+    }
+    ${optionalString (cfg.secretSettingsFile != null) "source ${cfg.secretSettingsFile}"}
+      exec ${cfg.package}/bin/funkwhale "$@"
+  '';
```
This currently gets installed into `/var/lib/funkwhale/bin/funkwhale`. It is required to import media as far as I can tell, so I wonder if we can somehow get it into PATH from inside this module.

SuperSandro2000 replied:
I think we would need to add it to systemPackages.

mweinelt committed:
python3Packages.pydub: patch ffmpeg paths

This allows pydub to find ffmpeg at runtime.

mweinelt commented:
The tests are are failing due to a core dump in python, from what I can tell around pillow. :facepalm:

mweinelt committed:
funkwhale: 1.2.1 -> 1.2.2

https://dev.funkwhale.audio/funkwhale/funkwhale/-/blob/1.2.2/CHANGELOG\#L13

Pins uvicorn to stay compatible with the asgiref pin, that is still
unresolved.

mweinelt commented:
Updated to 1.2.2. Needed to pin uvicorn to stay compatible with asgiref. Fixed django-auth-ldap (#160012) which got broken in the latest python-updates cycle.

And still gunicorn is crashing in the NixOS test. Would love for more eyes on this.

RaitoBezarius commented on pkgs/servers/web-apps/funkwhale/update.sh:
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env nix-shell
+#! nix-shell -i oil -p jq sd nix-prefetch-git ripgrep moreutils oil nixUnstable prefetch-yarn-deps
```
is there a need for ripgrep?

mweinelt replied:
No, thanks for noticing.

RaitoBezarius reviewed (approved):
It really looks nice.

RaitoBezarius commented:
> Updated to 1.2.2. Needed to pin uvicorn to stay compatible with asgiref. Fixed django-auth-ldap (#160012) which got broken in the latest python-updates cycle.
> 
> And still gunicorn is crashing in the NixOS test. Would love for more eyes on this.

I will take a look when I have a chance.

RaitoBezarius commented:
At the moment, I'm getting something like:
```
$ nix --experimental-features nix-command build --no-link --keep-going --option build-use-sandbox relaxed -f /home/raito/.cache/nixpkgs-review/pr-152160/build.nix
error: build of '/nix/store/mr61ri5d703pm9fxpjn0i4fv1v8qlm9a-python3.9-azure-core-1.21.1.drv' on 'ssh://dc1-playground-builder' failed: error: builder for '/nix/store/mr61ri5d703pm9fxpjn0i4fv1v8qlm9a-python3.9-azure-core-1.21.1.drv' failed with exit code 1;
              last 10 log lines:
              > ERROR tests/test_settings.py::TestStandardSettings::test_setting_exists[log_level]
              > ERROR tests/test_settings.py::TestStandardSettings::test_setting_exists[tracing_enabled]
              > ERROR tests/test_settings.py::TestStandardSettings::test_setting_env_var[log_level]
              > ERROR tests/test_settings.py::TestStandardSettings::test_setting_env_var[tracing_enabled]
              > ERROR tests/test_settings.py::TestStandardSettings::test_init - ValueError: D...
              > ERROR tests/test_settings.py::TestStandardSettings::test_config - ValueError:...
              > ERROR tests/test_settings.py::TestStandardSettings::test_defaults - ValueErro...
              > ERROR tests/test_settings.py::TestStandardSettings::test_current - ValueError...
              > ERROR tests/test_universal_pipeline.py::test_pipeline_context - ValueError: D...
              > ========= 53 passed, 838 deselected, 7 warnings, 178 errors in 46.78s ==========
              For full logs, run 'nix log /nix/store/mr61ri5d703pm9fxpjn0i4fv1v8qlm9a-python3.9-azure-core-1.21.1.drv'.
error: builder for '/nix/store/mr61ri5d703pm9fxpjn0i4fv1v8qlm9a-python3.9-azure-core-1.21.1.drv' failed with exit code 1;
       last 10 log lines:
       > ERROR tests/test_settings.py::TestStandardSettings::test_setting_exists[log_level]
       > ERROR tests/test_settings.py::TestStandardSettings::test_setting_exists[tracing_enabled]
       > ERROR tests/test_settings.py::TestStandardSettings::test_setting_env_var[log_level]
       > ERROR tests/test_settings.py::TestStandardSettings::test_setting_env_var[tracing_enabled]
       > ERROR tests/test_settings.py::TestStandardSettings::test_init - ValueError: D...
       > ERROR tests/test_settings.py::TestStandardSettings::test_config - ValueError:...
       > ERROR tests/test_settings.py::TestStandardSettings::test_defaults - ValueErro...
       > ERROR tests/test_settings.py::TestStandardSettings::test_current - ValueError...
       > ERROR tests/test_universal_pipeline.py::test_pipeline_context - ValueError: D...
       > ========= 53 passed, 838 deselected, 7 warnings, 178 errors in 46.78s ==========
       For full logs, run 'nix log /nix/store/mr61ri5d703pm9fxpjn0i4fv1v8qlm9a-python3.9-azure-core-1.21.1.drv'.
error: 1 dependencies of derivation '/nix/store/xpyp12r4wy8hfk6f2ivhq72y4xdyijfr-python3.9-azure-servicebus-7.5.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/635cvg56hzciyaqwr1yihscmxa12s5wp-python3.9-kombu-5.2.4.drv' failed to build
error: build of '/nix/store/6b19zxn95i5h2rys6q344knv54qfr6c9-tensorflow-2.7.1.drv' on 'ssh://dc1-playground-builder' failed: error: builder for '/nix/store/6b19zxn95i5h2rys6q344knv54qfr6c9-tensorflow-2.7.1.drv' failed with exit code 1;
              last 10 log lines:
              > [3,369 / 5,838] Compiling tensorflow/core/kernels/cwise_op_div.cc; 467s local ... (32 actions running)
              > [3,369 / 5,838] Compiling tensorflow/core/kernels/cwise_op_div.cc; 1167s local ... (32 actions running)
              > ERROR: /build/source/tensorflow/compiler/mlir/tensorflow/BUILD:436:15: C++ compilation of rule '//tensorflow/compiler/mlir/tensorflow:tensorflow_ops_n_z' failed (Exit 1): gcc failed: error executing command /nix/store/pja9g36cy32z3d51942jqk91a6l2d5nv-gcc-wrapper-10.3.0/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 ... (remaining 136 argument(s) skipped)
              > gcc: fatal error: Killed signal terminated program cc1plus
              > compilation terminated.
              > INFO: Elapsed time: 2337.271s, Critical Path: 1211.43s
              > INFO: 3430 processes: 691 internal, 2739 local.
              > FAILED: Build did NOT complete successfully
              > FAILED: Build did NOT complete successfully
              > 
              For full logs, run 'nix log /nix/store/6b19zxn95i5h2rys6q344knv54qfr6c9-tensorflow-2.7.1.drv'.
error: 1 dependencies of derivation '/nix/store/4hnzmg4p0rrhx04pb3r1qkyg327xrijk-python3.9-celery-5.2.3.drv' failed to build
error: 2 dependencies of derivation '/nix/store/zbd00q5lwc6k6l8b65hd8ry37fcy5aa9-funkwhale-1.2.2.drv' failed to build
error: builder for '/nix/store/6b19zxn95i5h2rys6q344knv54qfr6c9-tensorflow-2.7.1.drv' failed with exit code 1;
       last 10 log lines:
       > [3,369 / 5,838] Compiling tensorflow/core/kernels/cwise_op_div.cc; 467s local ... (32 actions running)
       > [3,369 / 5,838] Compiling tensorflow/core/kernels/cwise_op_div.cc; 1167s local ... (32 actions running)
       > ERROR: /build/source/tensorflow/compiler/mlir/tensorflow/BUILD:436:15: C++ compilation of rule '//tensorflow/compiler/mlir/tensorflow:tensorflow_ops_n_z' failed (Exit 1): gcc failed: error executing command /nix/store/pja9g36cy32z3d51942jqk91a6l2d5nv-gcc-wrapper-10.3.0/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 ... (remaining 136 argument(s) skipped)
       > gcc: fatal error: Killed signal terminated program cc1plus
       > compilation terminated.
       > INFO: Elapsed time: 2337.271s, Critical Path: 1211.43s
       > INFO: 3430 processes: 691 internal, 2739 local.
       > FAILED: Build did NOT complete successfully
       > FAILED: Build did NOT complete successfully
       >
       For full logs, run 'nix log /nix/store/6b19zxn95i5h2rys6q344knv54qfr6c9-tensorflow-2.7.1.drv'.
error: 1 dependencies of derivation '/nix/store/dqjgh358vp6cl9hb2f0rnxy8ykwqj9g9-python3.9-tensorflow-2.7.1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/rkvfw69vzk77wxx9pmn32b3aipwx3pr4-python3.9-tensorflow-datasets-4.4.0.drv' failed to build
error: 2 dependencies of derivation '/nix/store/bxd2ix73im2m9ghpnsrfjdqpayjqky2z-review-shell.drv' failed to build
``` while reviewing, did you encounter these errors @mweinelt ? Or is it just my machine who is not powerful enough for TensorFlow?

mweinelt commented:
Yeah, some python packages are broken on master right now, annoying. Alot of the required fixes are in #162538.

RaitoBezarius commented:
> Yeah, some python packages are broken on master right now, annoying. Alot of the required fixes are in #162538.

Alright, will revisit this when things are better :).

mweinelt committed:
python3Packages.django-cache-memoize: init at 0.1.10

SuperSandro2000 commented on pkgs/servers/web-apps/funkwhale/update:
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i oil -p jq sd nix-prefetch-git moreutils oil nixUnstable prefetch-yarn-deps
```
```suggestion
#!nix-shell -i oil -p jq sd nix-prefetch-git moreutils oil nix prefetch-yarn-deps
```

SuperSandro2000 commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,378 @@
+          alias = cfg.package.frontend + "/";
+          extraConfig = ''
+            add_header Cache-Control "public, must-revalidate, proxy-revalidate";
+            add_header Content-Security-Policy "default-src 'self'; script-src 'self'; style-src 'self' 'unsafe-inline'; img-src 'self' data:; font-src 'self' data:; object-src 'none'; media-src 'self' data:; worker-src 'self'";
```
```suggestion
            add_header Content-Security-Policy "default-src 'self'; style-src 'self' 'unsafe-inline'; img-src 'self' data:; font-src 'self' data:; object-src 'none'; media-src 'self' data:;";
```

SuperSandro2000 commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,378 @@
+    };
+
+
+    services.nginx = {
+      enable = mkDefault true;
```
We should add an option wether nginx should be enabled which also disables the other options if people want to do with another reverse proxy server.

SuperSandro2000 commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,378 @@
+        '';
+        serviceConfig = {
+          Type = "oneshot";
+        } // commonServiceConfig;
+      } // commonUnitConfig;
```
I'd imagine we could replace this with a single lib.recursiveUpdate

SuperSandro2000 commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,378 @@
+          mkdir -p $STATE_DIRECTORY/bin
+          ln -sf ${manage} $STATE_DIRECTORY/bin/funkwhale
+
+          ${cfg.package}/bin/funkwhale migrate
+          ${cfg.package}/bin/funkwhale collectstatic --no-input
```
```suggestion
          ${cfg.package}/bin/funkwhale migrate
          ${cfg.package}/bin/funkwhale collectstatic --no-input
```
Would it work to add those to path?

SuperSandro2000 commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,378 @@
+            type = types.ints.unsigned;
+            default = 0;
+            example = 4;
+            description = "Number of concurrent tasks that a single celery worker can execute.";
```
0 means it is automatically choosen?

SuperSandro2000 commented on nixos/modules/services/web-apps/funkwhale.nix:
```diff
@@ -0,0 +1,378 @@
+
+          FUNKWHALE_PROTOCOL = mkOption {
+            type = types.enum [ "http" "https" ];
+            default = "http";
+            example = "https";
```
I kinda want to default to https.
